pkgs/_macro_client/lib/macro_client.dart

```diff
@@ -36,21 +43,54 @@ class MacroClient {
     return MacroClient._(macros, socket);
   }
 
-  void _send(Map<String, Object?> node) {
+  void _sendRequest(MacroRequest request) {
+    // TODO(davidmorgan): currently this is JSON with one request per line,
+    // switch to binary.
+    socket.writeln(json.encode(request.node));
+  }
+
+  void _sendResponse(Response response) {
     // TODO(davidmorgan): currently this is JSON with one request per line,
     // switch to binary.
-    socket.writeln(json.encode(node));
+    socket.writeln(json.encode(response.node));
   }
 
   void _handleRequest(String request) async {
-    // TODO(davidmorgan): support more than one request type.
-    final augmentRequest =
-        AugmentRequest.fromJson(json.decode(request) as Map<String, Object?>);
-    // TODO(davidmorgan): support multiple macros.
-    final response = await macros.single.augment(host, augmentRequest);
-    _send(response.node);
+    final jsonData = json.decode(request) as Map<String, Object?>;
+    final hostRequest = HostRequest.fromJson(jsonData);
+    switch (hostRequest.type) {
+      case HostRequestType.augmentRequest:
+        _sendResponse(Response.augmentResponse(
+            await macros.single.augment(host, hostRequest.asAugmentRequest)));
+      default:
+      // Ignore unknown request.
+    }
+    final response = Response.fromJson(jsonData);
+    switch (response.type) {
+      case ResponseType.unknown:
+        // Ignore unknown response.
+        break;
+      default:
+        // TODO(davidmorgan): track requests and responses properly.
+        if (_responseCompleter != null) {
+          _responseCompleter!.complete(response);
+          _responseCompleter = null;
+        }
+    }
   }
 }
 
 /// [Host] that is connected to a remote macro host.
-class RemoteMacroHost implements Host {}
+class RemoteMacroHost implements Host {
```

Comment:
As long as the client only sees the more restricted interface I think it would be fine? Yes, you are correct people could use `dynamic` to reach into the details of the class though. But at least right now I am not concerned about that.

Reply:
Yes, it's only needed from a security point of view.

For our own benefit I think it's nice to start splitting into client implementation details vs what is provided to user code, although it would be nicer still without the "this" wrapper :) probably it ends up moving out and being a clearer boundary.